Adds article about testing speed (#76)
thefrontside/frontside.com: src/pages/blog/2020-triple-threat-to-testing-part-1-speed.md

---
templateKey: blog-post
title: >-
  The Triple Threat to Testing — Part 1: Speed
date: 2020-03-31T05:00:00.000Z
author: Charles Lowell
description: >-
  More and more teams are using Bluetooth for innovative products, but developing UIs for connected devices can be challenging: limited physical devices slow down development and make it nearly impossible to keep up a strong test suite. Our solution is simulation and Open Source power.  
tags:
  - testing
image: /img/2020-triple-threat-to-testing-part-1-speed.png
---
![The Triple Threat to Testing — Part 1\:Speed](/img/2020-triple-threat-to-testing-part-1-speed.png)

> Writing tests is like paying taxes. And much like how attitudes vary towards paying taxes, attitudes diverge widely in development circles about writing tests. Some people literally dread paying their taxes, and similarly there are developers who would rather cut off their arm than spend time testing an app. Other people willingly pay their taxes because of the benefits they confer, like drivable roads, health care and national defense. In similar fashion there are those developers that look forward to testing because they can provide confidence while protecting against regressions. There’s even those people who argue that there’s even something inherently good in paying taxes and writing tests—a moral obligation if you will.

Reviewer: ```suggestion
Writing tests is like paying taxes. And much like how attitudes vary towards paying taxes, attitudes diverge widely in development circles about writing tests. Some people dread paying their taxes; much in the same way as some developers would rather cut off their arm than spend time testing an app. Other people willingly pay their taxes because of the benefits they confer, like drivable roads, health care and national defense. In similar fashion there are those developers that look forward to testing because of the protection in gives against regressanios and the confidence it inspires when making changes. You'll even find people who argue that there’s something inherently good in paying taxes and writing tests—a moral obligation if you will.
```

I feel like this flows better

But just like people debate the merits of taxes, developers argue over the value of tests. Some claim that they don’t need tests while others think they are so important you should write them before anything else. There’s no consensus in part because tests vary widely across technologies, projects, teams, and organizations. One thing is for certain, however; just like taxes take money out of your pocket, testing imposes a variety of costs on your organization and your development team. 

> We’re not writing to convince you about the value of testing any more than we’re here to convince you to pay your taxes. We instead want to share our thinking about three essential make-or-break factors in tests that dramatically affect how much of a burden they pose. The first factor is the most obvious: speed. Slow tests are literally a drag on teams and organizations, but we want to share some ideas that can help improve testing speed.

Reviewer: I don't think our goal is to share thinking about make-or-break factors, or rather, it isn't our primary goal. I think we do want to convince them primarily that there is a difference in what you get for your tax money. Tax money can be wasted, and can also be well spent, and that what you really want is _good governance_. A structure that yields the highest return on the investment of your taxes.

```suggestion
We’re not writing to convince you about the value of testing any more than we’re here to convince you to pay your taxes. But I think we can all agree, that if you _are_ going to pay taxes, then you should expect the _best governance_ in return. So when it comes to testing, our experience has show us that if you _are_ going to invest in tests, there are three essential make-or-break factors in that will dramatically affect how much benefit they will bring in exchange for how much burden they pose. 

The first factor is the most obvious: speed. Slow tests are literally a drag on teams and organizations.
```

## Why Speed Matters

In the stock market, information services offer different pricing schemes around speed. You can wait and get information for free in a minute or pay to get data within seconds. But if you want information with minimum latency the price tag goes through the roof. Why? The infrastructure to make this possible is more expensive and the value of knowing something before others is very high. 

> Speed gives brokers a competitive advantage, and the same is true when it comes to testing. How rapidly a test can diagnose issues in your application can make all the difference. Because tests determine the minimum rhythm at which developers can discuss the results of their code, the speed of a test suite scales up to affect the entire organization. A Pull Request whose test suite takes 1 hour to run winds up setting a plodding pace not just for the development team but product and design teams as well.

Reviewer: ```suggestion
Speed gives brokers a competitive advantage, and the same is true when it comes to testing. How rapidly a test can diagnose issues in your application can make all the difference. Because tests determine the minimum rhythm at which developers can discuss the results of their code, the speed of a test suite scales up to affect the entire organization. A Pull Request whose test suite takes 1 hour to run sets a plodding pace not just for the development team but product and design teams as well.
```

Reviewer: "winds up setting" was a bit verbose and broke up the flow for me.

Reviewer 2: I like sets too

What’s worse is that when tests don’t deliver results fast enough, a developer working on a feature loses the thread of the problem the test was intended to help solve. This dynamic quickly leads to frustration and causes developers—who are inherently problem-solvers—to find work-arounds that essentially avoid testing altogether. Whatever you think about testing, we can all agree that a test suite should be fast enough not to break the context under which it’s being run. An unused test suite isn’t a test suite at all.

> With this level of disfunction you’d think that improving test speed would be at the top of everyone’s to-do list. Yet if you ask around it seems like we have just accepted that tests are slow and have learned to adapt to their plodding pace. Some teams decide to run only a few tests on every PR through CI, reserving the whole suite for ‘important’ builds. Other teams choose from time to time to not even bother running tests locally. 

Reviewer: ```suggestion
With this level of dysfunction you’d think that improving test speed would be at the top of everyone’s to-do list. Yet if you ask around it seems like we have just accepted that tests are slow and have learned to adapt to their plodding pace. Some teams decide to run only a few tests on every PR through CI, reserving the whole suite for ‘important’ builds. Other teams choose from time to time to not even bother running tests locally. 
```

Reviewer: While both are correct, "dysfunction" is the more common variant

> These reactions are typical and even make sense: waiting for a test suite to run is boring and non-productive, like standing in line at the Department of Motor Vehicles. What’s worse is that you have to go to the back of the line time and again when the scripts you are running produce results that are flaky. It’s depressing to realize that as a result of testing the development process—which at its core is an essentially creative process—has come to feel bureaucratic.

Reviewer: ```suggestion
These reactions are typical and even make sense: waiting for a test suite to run is boring and non-productive, like standing in line at the Department of Motor Vehicles. What’s worse is that you have to go to the back of the line time and time again when the scripts you are running produce results that are flaky. It’s depressing when the development process—which at its core is an essentially creative process—has come to feel bureaucratic as a result of testing.
```

I re-worded the last sentence because I couldn't stop reading "testing the development process" as a single activity, when really they are two stepparent things. 

I feel like we should make a call back to our initial thesis which is that you need to get value out of your taxes, and if you're in this situation, then that's not what you're getting in exchange

### "Slow" Front-end Apps 

User interfaces are supposed to be used by humans, but when a machine uses them they reveal their underlying performance issues. A person furiously tapping your app buttons as fast as they can will not match the speed of a machine. Not even developers themselves – who know exactly where each button is located – can assert the state of the app as fast as a script. That’s why slow loading and reacting UI elements are invisible to us most of the time.

When you run a test the machine will always be limited by the speed of the application. On every simulated interaction and on every assertion, the test runner waits: it waits for code to be loaded and parsed, for elements to appear, for APIs to load, or for interactions to happen. Repetition scales up the natural latency of an app until it becomes noticeable. 

*Right now you can do the following in order to streamline front-end apps*: 

> - **Check the [Time to Interactive](https://developers.google.com/web/tools/lighthouse/audits/time-to-interactive) of your application** You may be loading too many things on the entry point of your app: user profile, dynamic navigation segments, analytics scripts, and more. Consider if there’s something you shouldn’t be using within the test context (like marketing pixel scripts).

Reviewer: ```suggestion
- **Check the [Time to Interactive](https://developers.google.com/web/tools/lighthouse/audits/time-to-interactive) of your application** You may be loading too many things on the entry point of your app: user profile, dynamic navigation segments, analytics scripts, and more. Consider if there’s something you shouldn’t be using within the test context (like pixel scripts for marketing).
```

again re-word, because "marketing pixel scripts" sounds like an activity in of itself.... "hey, you want to come help me market pixel scripts this afternoon?"

- **Disable non-test-related latency**: for example, unless you’re testing the animation itself, you can consider disabling animations in the testing context. Animations just mean waiting time for the test runner in most cases.

> - **Evaluate the overall performance of your app**: perhaps the test suite is revealing that your app has a performance problem that you have to address. [Google’s Lighthouse](https://developers.google.com/web/tools/lighthouse) can help identify key performance indicators, which you can also integrate into your development process.

Reviewer: When I read this, it didn't resonate with me. I was like "duh." Of course this is true.... and I was there when we came up with the idea to include it!

Should we help folks let it settle with something like "this seems obvious, but sometimes it pays to stop and contemplate what's right in front of you.... if the test are painfully slow, then perhaps your app is painfully slow, and an investment in making it more peppy will yield dividends for both your developers _and_ your users" or something like that.

Author: I think this is great, I'll add it

### Slow Build times

> The first step of a test suite usually involves generating a fresh build of your app in the context that you choose. Although this is a time cost that the test suite generally only has to pay once, it is still a major factor in how soon developers can get test results. A fresh build can take up several minutes – the equivalent of eons in computer time.

Reviewer: ```suggestion
The first step of a test suite usually involves generating a fresh build of your app in the context that you choose. Although this is a time cost that the test suite generally only has to pay once, it is still a major factor in how soon developers can get test results. A fresh build can take up to several minutes – the equivalent of eons in computer time.
```

*Right now you can do the following in order to speed up build times*:

- **Budget for performance improvements**: build times tend to be a boiling problem: as features are constantly added, the time increases without anybody noticing. Performance improvements are typically triggered only after somebody complains about the tools being too slow. We recommend having a performance budget to proactively improve the build times; otherwise you might find yourself returning to re-write the build script again and again.
- **Evaluate your test environment**: check with your team if you can simulate parts of the test environment (for instance, use a tool like [Mirage.js](https://miragejs.com/) instead of using a real database).

### Not specific enough entry points

> Web applications are attractive because you don’t need to install anything to use them: you simply hit a URL and you’re ready to do what you want. The URL is the entry point for a web application. Most people access the app through the root `/` and then navigate their way around. Advanced users may even know the URLs they care about and go straight to it (`/p/{my-crush}` or even `/p/{my-crush}/photos`). Using more specific entry points saves time of navigation to users, and can save time when testing. 

Reviewer: ```suggestion
Web applications are attractive because you don’t need to install anything to use them: you just hit a URL and you’re ready to do what you want. The URL is the entry point for a web application. Most people access the app through the root `/` and then navigate their way around. Advanced users may even know the URLs they care about and go straight to it (`/p/{my-crush}` or even `/p/{my-crush}/photos`). Using more specific entry points saves time of navigation to users, and can save time when testing. 
```

Overuse of "simply" is one of my pet peeves

- **Simulate authentication**: discuss with your team the best strategy to simulate authentication in your application, from using a framework specific tool to manipulating local storage directly (while not forgetting to take security considerations into account).

### Waiting too much, waiting too little

Asserting the state of a web application is like playing the street basketball game 21. The first team to score exactly 21 points wins; but if they go over their score is reduced by 11. Assertions are similar in that you must make them at a specific moment in time: if you don’t hurry up you lose precious time, but if you miss the mark you’re stuck with more waiting and have to try again. 

Web applications perform a lot of tasks asynchronously: loading files, fetching images, calling APIs and recording analytics to name a few. That’s why waiting is such a tricky part of testing user interfaces. There’s no exact fixed time intervals in the way elements are rendered because they depend on many other elements and which client is being used. There’s also external dependencies like server requests that may take variable amounts of time.

> Sometimes developers can set tests to wait too much. For example, to make sure an element was rendered in the page we can make our test wait for a fixed time interval like `wait(‘500ms’)`. This tells the runner to halt the test and wait (often aimlessly). Other times developers can waste time setting the test to wait on the wrong element. 

Reviewer: I think this could use more analysis to quantify the problem and drive the point home. For example, if the element appears after 25ms, then I've been wasting 475ms! That's a ton! If I do that or something like that even 10 times during the course of my entire test suite (and very often its more than that), then I'm bleeding seconds and minutes each run, which turn into minutes and hours over weeks, which durn into days and weeks over months.....

So while it's true that the wait is "aimless", it's more important to describe the carnage that aimless waits can have on your team productivity as they scale up.

Author: We can add some quantifying, but I don't want to spend too many words on it nor make it sound overly _dramatic_. I think the trend of micro-optimizations and its decline has left that kind of progressive sum of little times in a not entirely relevant position.

The most significant advantage of getting a test environment closer to production is that you gain confidence that your tests are diagnosing your application realistically. However, running a test suite against real APIs represents a major challenge in terms of time and cohesiveness. The same can be said of other external dependencies (like Bluetooth connectivity) for your application. 

*Right now you can do the following in order to select the right testing scenario:*

- **Assess the level of realness your team needs**: perhaps your data needs are not too complex, and your application doesn’t handle critical operations, so mocks would suffice. If your business requires more sophisticated data management, you may need to invest in tools that can help you ensure your tests cover such cases.
- **Consider simulation**: simulation is an intermediate point between mocks and real APIs that provides the velocity of mocks with the complexity of real backends. You can simulate database queries, as well as errors, using tools like [Mirage.js](https://miragejs.com/).

<hr><br>

> Everybody feels better about paying taxes when they can see results. At Frontside we think everyone would feel better about tests if they could see their results faster. By addressing both application concerns (time to interactive, build time, routing granularity) and testing strategies (waiting precision, realness) developers right now can win back precious seconds, minutes, and even hours of time spent on slow tests. We think it’s worth the effort to improve testing, as speedy test suites shorten feedback loops and add up to increase the organization’s velocity. 

Reviewer: I feel like the connection to the foundational analogy (paying taxes) is not as strong as it could be. Maybe instead of starting the conclusion with that connection, we can end with it?

Author: I think we can improve the conclusion in that case. It's usually nice to end a text with a parallel of how you started it. What can we say in terms of taxes that applies to speed?

Reviewer: Maybe just adding something like "We think it’s worth the effort to improve testing, as speedy test suites shorten feedback loops and add up to increase the organization’s velocity. And that maximizes the dividend you get for your investment in tests."

> Next we want to talk about the second essential make-or-break factor in testing: reliability. Tests can be flaky, but there are strategies you can take now to detect, fix, and even prevent having non-deterministic tests in your application.

Reviewer: The transition between this and the preceding paragraph is too abrupt. We need something like a small heading:

### Up next....

Speed is critical to a healthy testing strategy, but it won't work without the other pillars of the testing triple threat. In part two, we'll explore the second essential make-or-break.....

Author: Yes, this is great. I'll add it!

---
templateKey: blog-post
title: >-
  The Triple Threat to Testing — Part 1: Speed
date: 2020-03-31T05:00:00.000Z

> author: Charles Lowell

Reviewer 3: What should we do for this field?

Author: I will pair with Taras to find a solution so we can pull in more than one author. If it's not possible, my plan was to release the next in the series in the name of other people.

description: >-

>   More and more teams are using Bluetooth for innovative products, but developing UIs for connected devices can be challenging: limited physical devices slow down development and make it nearly impossible to keep up a strong test suite. Our solution is simulation and Open Source power.  

Reviewer 3: We'll need to come up with a little description here.

Author: Wow good catch, thanks! I'll prepare a little description for this

The most significant advantage of getting a test environment closer to production is that you gain confidence that your tests are diagnosing your application realistically. However, running a test suite against real APIs represents a major challenge in terms of time and cohesiveness. The same can be said of other external dependencies (like Bluetooth connectivity) for your application. 

*Right now you can do the following in order to select the right testing scenario:*

- **Assess the level of realness your team needs**: perhaps your data needs are not too complex, and your application doesn’t handle critical operations, so mocks would suffice. If your business requires more sophisticated data management, you may need to invest in tools that can help you ensure your tests cover such cases.
- **Consider simulation**: simulation is an intermediate point between mocks and real APIs that provides the velocity of mocks with the complexity of real backends. You can simulate database queries, as well as errors, using tools like [Mirage.js](https://miragejs.com/).

> <hr><br>

Reviewer 3: Another spot where we can probably use a markdown HR.

```suggestion
---
```

Everybody feels better about paying taxes when they can see results. At Frontside we think everyone would feel better about tests if they could see their results faster. By addressing both application concerns (time to interactive, build time, routing granularity) and testing strategies (waiting precision, realness) developers right now can win back precious seconds, minutes, and even hours of time spent on slow tests. We think it’s worth the effort to improve testing, as speedy test suites shorten feedback loops and add up to increase the organization’s velocity. 

### Up next...

Speed is critical to a healthy testing strategy, but it won't work without the other pillars of the testing triple threat. Next we want to talk about the second essential make-or-break factor in testing: reliability. Tests can be flaky, but there are strategies you can take now to detect, fix, and even prevent having non-deterministic tests in your application.

> <hr><br>

Reviewer 3: Can we do a markdown HR? It's three or more hyphens:

`---`

👇

---

```suggestion
---
```
